Browse mockup : just add ui components

sushmaanandakoju opened the pull request:
_Purpose:_ PR attempts to build ui-components required to create a Browse page similar to HCA Browse page. Task was started prior to sprint and awaiting feedback on: https://docs.google.com/document/d/1rWJMt5Wx617OZ-U6HZf2JTBQwc1XuiZ43NZ9H28Zu_Q/edit 

As discussed with Chuck, this change, which was initially large, is split into smaller number of changes/branches for review. This is the first PR among 3 expected PRs.
Remaining files that were modified/changed for browse page mockup, will be added to another PR shortly.

_This PR consists of following:_

1) I tested various material-ui dropdowns/menus, suggestions and finally have retained that which are very basic in style, and implementation.
2) packages added: material-ui-chip-input, react-autosuggest, autosuggest-highlight, pubsub-js and flat.
3) Add search-facets, experiments json files similar to HCA browse page.
4) Mockup BaseDropDown, and SubMenuOptions together when populated from json would be: 
![Screenshot_20190926_141629](https://user-images.githubusercontent.com/8979477/65714082-6bb8bb00-e068-11e9-85c3-da8615b214bd.png)

5) Add Publisher/Subscriber pubsub-js implementation to subscribe to dropdown menu changes and maintaining the state of selected options. Add constants for topics subscriptions.
6) Add SearchInput to type and search. Add ChipInputs and Browse Summary to display browse summary of all selected inputs (from dropdowns, as well as from typed inputs). Would appear as :
![Screenshot_20190926_143214](https://user-images.githubusercontent.com/8979477/65715158-8db33d00-e06a-11e9-8391-740b9b1f263b.png)

7) Add AutoSuggest to show selections from search facets json file (as per material-ui-chip-input examples).

_W.I.P, Known issues:_
1) Search Input to update typed inputs when typed chips are removed from browseSummary.
2) Add more to BrowseSummary: such as number of experiments, number of tissues, cell counts etc that which is interest.
3) Browse Summary must be readble, editable but should not look like SearchInput - W.I.P .
3) AutoSuggest styling or simplifying the design : is ongoing.

_Known assumptions/points in discussion:_
1. We attempt to implement a browse page that is similar to that of HCA browse page.
2. These changes are not complete. But current state of ui-components needs review and merge.
3. Unsure, whether we receive feedback on UI after merge / before merge: since this is an ongoing-discussion about approaches to designing/implementing UI.
4. This is not a final UI design or even reviewed design. We are also simultaneously reviewing and working on making 4dn Fourfront UI re-usable - which is a w.i.p.

Commit:
Merge branch 'develop' of https://github.com/hubmapconsortium/hubmap-data-portal into develop

Commit:
add browse data components templates:

1) add dropdowns (Donor, Search, Specimen...)
2) add Menus template
3) add ExperimentsTable template

Commit:
add BrowseData page

Commit:
allow this {...props}

Commit:
change order of display of Browse menu

1) move experiments to first in menu list

Commit:
update renamed component name

Commit:
changes to shared ui components:

1) move ui components to ui-components folder
2) add dropdown menu, menu item, support class level pop over items
3) add chip component to show selected items from all menus on browse
4) add custom search input
5) add mimic flat ui styling (similar to material design)
6) do remove Browsedata, Menu and Chip from browse and pages folders

Commit:
modify donor dropdown menu:

1) add Donor Dropdown menu with props
2) add mock data to show in menu
3) add custom styling to DropdownMenu components

Commit:
change class components to functional components

1) modify SearchDropDown and SpecimenDropdown to function components
2) add styling and add default return to SearchDropdown

Commit:
modify Experiments table to include the Grid containers and components

1) add all ExperimentsTable, dropdown, search options

Commit:
rename ExperimentsTable

1) modify ExperimentsTable
2) remove page heading: Title.js from the Experiments Table

Commit:
export Browse UI components in index.js

Commit:
add Browse page title component

Commit:
add ControlledChip search and select form control dropdown

Commit:
add clsx for formcontrol select dropdown

Commit:
remove seach from Nav header

Commit:
add clsx for formcontrol select dropdown

Commit:
Revert "add clsx for formcontrol select dropdown"

This reverts commit 397b0c0ca76c018348cac24bc4825083440d0b6e.

Commit:
update dropdown ui controls

Commit:
add dropdown menu to specimen dropdown (like hca)

Commit:
add style, menu click and close events to dropdown menu

Commit:
add new ui controls : selectDropdown and controlledchip search controls
to exports

Commit:
update Experiments page:

1) remove styling for grid items
2) add dropdown menus
3) add extra grid items [replace later]

Commit:
eslint spacing: Experiments.js

Commit:
[w.i.p] add menu dropdown panel

Commit:
[w.i.p] modify DropdownMenu :
1) add checkboxes
2) test ui and alignment

Commit:
add new ui controls to export

Commit:
add facet search sample data for filling in dropdown menus

Commit:
ad package material-ui-chip-input

Commit:
[w.i.p] add re-usable BaseDropDownPanel

Commit:
add BaseDropDownpanel and add facets from json file (example hca )

Commit:
Add ElevationScroll and add baseDropdownPanel to main page:

1) remove DropDownMenu from experiments page
2) add BaseDropDownPanel in experiments page
3) Add ElevationScroll in ui-components

Commit:
Add baseChildDropDown

Add base child dropdown component to add multiple children to dropdown menu panel

Commit:
Add child dropdown menu to add as many as menusections props received from props

Commit:
Add BaseDropDown Menu with props from json

no need to pass the props onto children (BaseDropDown passes on to the base dropdown menu children)

Commit:
remove baseDropDown
now that DonorDropdown menus adds respective BaseDropDown menu panel

Commit:
add ElevationScroll to ui-components/index.js exports

Commit:
adjust to fit all subsections in single menu dropdown

Commit:
add package pubsub-js

Commit:
Add CustomSelectMenu with BaseChildDropDown checkboxes for search facets json

Commit:
add search/selected chips to search SelectedMenuSummary

Commit:
add PubSub in middleware

1) Add custom pub/sub methods
2) add exports in middleware

Commit:
Add FileDropdownMenu from search facet json

Commit:
Add method Dropdown menu from search facets json

Commit:
Add Tissuetype dropdown from search facets json

Commit:
Add pubsub implementation:

1) modify style for FormLabel
2) update propTypes
3) Add publish selected menus options

Commit:
add 0 margin for parent dropdown and 100% width
note: not using now (may remove later, used for testing ONLY) .

Commit:
Change to custom select dropdown menu

Commit:
add custome select dropdown menu

Commit:
add Proptypes to the chips

Commit:
remove duplicate key ("unspecified")

Commit:
add SELECTED_MENU_ITEMS for pubsub global topic name

Commit:
add newly added Dropdown menus to exports in browse/

Commit:
add newly added select menu to exports in ui-components/

Commit:
add new dropdown menus, adjust styling to show in single line

Commit:
Add index.js for commons/ for all exports in commons
TODO: update al other individual file references later

Commit:
add TYPED_SEARCH_OPTIONS constant

1) add to constants for publising to topic
2) so TODO: later use this topic subscription for fetching from elasticsearch api

Commit:
Update Search Dropdown to use Controlled chipinput that has custom themel/layout

Commit:
remove additional publishers, subscribers list
1) PubSub-js maintains these lists
2) no need to duplicate mainatining own pubs/subs lists

Commit:
remove additional grid spaces and add summary option

1) Add SelectedMenuSummary
2) remove additional grid spaces

Commit:
update renamed SelectedBrowseOptionsSummary

Commit:
rename, add changes to subscribe to pubslished topics for selected/typed options

1) rename the file
2) add subscriptions to typed search text and selected menus from dropdown

Commit:
remove old file (after renamed file SelectedBrowseOptionsSummary)

Commit:
changes for publishing typed inputs

1) add only 1 chip at a time
2) remove one chip at a time
3) do update the chips array after changes
4) root theme MuiOutlinedInput - updated [w.i.p]

Commit:
changes to remove unused lines of code, add specific-chip when publishing

1) do add menu name the option is selected from
2) removed unused dict, console statements
3) adding menuname -> identified elastic search filters - facets structures

Commit:
update renamed file/export in index.js for ui-components

Commit:
update to material-ui 4.4.2 version

1) has bug fixes for chips, MuiOutlinedInput

Commit:
remove commented line from styles

Commit:
add grey color

Commit:
add MUIThemeProvider for on focus states

Commit:
remove unused statement /spaces

Commit:
save state of dropdowns
1) childdropdown to get state from parent
2) parent shoudl save state of child dropdown with multiple checkbox panels
3) parent subscribe to changes for child menu with checkboxes

Commit:
add checked/unchecked (topics)

Commit:
change topic name (renamed to CHECKED_MENU_OPTIONS)

Commit:
changes summary:

1) remove updating event.target.checked
2) add two events for CHECKED/UNCHECKED
3) update only single array
4) do check if the values exist before populating checked or not

Commit:
changes summary:

1) add events for CHECKED/UNCHECKED
2) update menu summary as per subscriptions

Commit:
check two states:
menu checked now and menu checked earlier + reopening the dropdown now

Commit:
update mock search facets json:

1) add Assays
2) update organs
3) update imaging file formats

reference: https://docs.google.com/spreadsheets/d/1Iki9vmENikmZTlv3UNkbcqts12_jlnCavXqCxpSQelY/edit#gid=1570824689

Commit:
add Assays dropdown list

1) Add AssaysDropdown
2) export dropdown in index.js

Commit:
Add Assaysdropdown to Grid

1) Add Assays dropdown to grid
2) Adjust spacing for all dropdowns
3) resize container grid to have optimal width

Commit:
add experiments json for mock data

reference: https://data.humancellatlas.org/explore/projects

Commit:
add experiments rows and columns headers from json

1) add column names, rows from experiments.json
2) adjust column sizes from json content (percentage sizes of overall width)
3) add styling to column header
4) add elements from json -> keys and iter

Commit:
modify experiments browse page for table content

1) add a divider after caption for experiments table
2) add table caption styling (underline, color etc)
3) adjust width of entire grid container

Commit:
add blue color on hover/focus in select dropdown and search inputs

reduce intensity of grey color for select dropdown when menu is open

Commit:
remove console.log and onOpen/onClose (not needed for now)

Commit:
space-eslint change

Commit:
add column sizes in percentage of container width

Commit:
eslint space fix

Commit:
dont disable input, but just make it uneditable (onkeydown, onPaste, onFocus )

Commit:
remove unused(ui components) components (not used anymore, after templating)

1) the dropdown panels/Chip components are replaced.
2) update exports in index.js

Commit:
add sorting to table [w.i.p]

1) TableSortLabel added to table

Commit:
add sorting and table head props[w.i.p]

Commit:
remove duplicated record from json

Commit:
add sorted entries based on column sort

1) add sorted rows
2) add TableSortedlabel
3) add custom Table head

Commit:
fix eslint error for AssaysDropdown menu

Commit:
fix eslint errors for SpecimenDropdown

Commit:
fix eslint errors for DonorDropdown

Commit:
fix eslint errors for FileDropdown

Commit:
fix eslint errors for MethodDropdown

Commit:
fix eslint errors for TissueTypedropdowm

Commit:
fix eslint errors for title

Commit:
fix eslint errors for ExperimentsTable

Commit:
fix eslint errors for SearchInput
rename SearchDropdown to SearchInput

Commit:
updated renamed file/export

Commit:
fix eslint errors

Commit:
fix eslint error for Experiments page

Commit:
fix eslint for BaseChildDropdown

Commit:
fix eslint ControlledChipInput

Commit:
fix eslint errors for CustomSearchInput

Commit:
fix eslint errors for CustomSelectMenu

Commit:
remove unused component: ElevationScroll

Commit:
fix eslint errors: SelectedBrowseOptionsSummary

Commit:
remove deleted file export

Commit:
updated renamed component: SearchInput

Commit:
add auto-suggest-highlight, react-autosuggest, flat packages

1) for auto-suggest in search inputs
2) for match and parse
3) deep flattening of json

Commit:
add packages for auto-suggest, suggestions

Commit:
Add AutoSuggest component

Commit:
export in index.js

Commit:
removing trailing comma in package.json

Commit:
Add Autosuggest component from (material-ui-input/examples )
render from json

Commit:
Merge branch 'develop' of https://github.com/hubmapconsortium/hubmap-data-portal into mock-data-browse

Commit:
have only ui-components/ , commons/ and middleware/ changes for PR

Commit:
add back missing file docker_util/__init__.py

Commit:
rename BaseSelectMenu, SubMenuOptions, SearchInput, BrowseSummary

1) rename files
2) update exports in index.js

Commit:
fix eslint errors for AutoSuggest file

mccalluc commented on hubmap/frontend/.eslintrc.yml:
```diff
@@ -19,3 +19,4 @@ rules:
   react/jsx-filename-extension: [0]  # JSX is ubiquitous; We don't need to make it explicit.
   react/sort-comp: [0]  # Non-alphabetical groupings can make more sense.
   react/jsx-one-expression-per-line: [0]  # Makes punctuation after tab awkward.
+  react/jsx-props-no-spreading: [0] # should be supported, for custom components
```
If you have particular components where prop spreading is intended, please list explicit exceptions: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-props-no-spreading.md#exceptions

sushmaanandakoju replied:
incorporated. added exceptions.

mccalluc commented on hubmap/frontend/package.json:
```diff
@@ -5,19 +5,24 @@
     "mui-datatables": "^2.4.0",
     "node-sass": "^4.12.0",
     "prop-types": "^15.7.2",
+    "pubsub-js": "^1.7.0",
```
I am very reluctant to add yet another way that state is managed, on top of plain react, and redux.

sushmaanandakoju replied:
Interesting. redux is something I'd be glad to use for like a middleware such as for wrapping api interactions. Even for middleware, PubSub is a handy package to send out some status. Given that we don't have an api yet, we should revisit to define what is required for "middleware" and for component-to-component interactions. We are not using any apis /redux for ui-components since everything is mockup from json.

Generally, I think selected options can filter multiple components: experiments list/table, summary, search and other pages also have similar functionality: on home page: if a search for gene expression started, would filter the charts, human anatomy map, summary, etc. So we should keep this as it is now and revisit later when we consider middleware design for react app. We always can **later**, if required, write custom pub-sub with custom global store rather than redux/pubssub-jsother popular packages. we do not have any active issues for middleware design in react right now, isn't it?

mccalluc commented on hubmap/frontend/src/commons/index.js:
```diff
@@ -0,0 +1,8 @@
+/** *
```
Why?
```suggestion
/**
```

mccalluc commented on hubmap/frontend/src/commons/index.js:
```diff
@@ -0,0 +1,8 @@
+/** *
+ * This exports all required components for browse data page.
```
If this is just for the browse page, perhaps name something other than `common`... or do you anticipate that they will be used more broadly?

sushmaanandakoju replied:
Yes, actually commons is used more broadly. which needs to be changed on a different PR to percolate changes for all other references to commons from all other components outside scope of this PR.

mccalluc commented on hubmap/frontend/src/components/.directory:
```diff
@@ -0,0 +1,4 @@
+[Dolphin]
+Timestamp=2019,9,26,13,48,27
+Version=4
+ViewMode=1
```
What is this? Should it be committed?

sushmaanandakoju replied:
removed.

mccalluc commented on hubmap/frontend/src/data/searchfacets.json:
```diff
@@ -0,0 +1,44 @@
+{
```
Please avoid just running words together: maybe just `facets.json`, or `searchFacets.json`, or `search-facets.json`.

mccalluc commented on hubmap/frontend/src/data/searchfacets.json:
```diff
@@ -0,0 +1,44 @@
+            "Sample": ["Unspecified", "end stage renal failure"]
+        },
+        "Assays":{
+            "scRNAseq":["snDropseq", "scRNAseq","IHC","LRET-IF","PER-DEI"],
```
```suggestion
            "scRNAseq": ["snDropseq", "scRNAseq", "IHC", "LRET-IF", "PER-DEI"],
```
Here and elsewhere, please whitespace appropriately. Your IDE may be able to help.

mccalluc commented on hubmap/frontend/src/middleware/PubSub.js:
```diff
@@ -0,0 +1,17 @@
+import PubSub from 'pubsub-js';
+
+export default class PubSubApi {
+tokens = {};
```
Body of class should be indented.

mccalluc commented on hubmap/frontend/src/middleware/PubSub.js:
```diff
@@ -0,0 +1,17 @@
+import PubSub from 'pubsub-js';
+
+export default class PubSubApi {
```
```suggestion
export default class {
```
With a default export, the internal name doesn't matter, so don't include it.

sushmaanandakoju replied:
We might want to add redux+pubsub at some point. but I removed it for now, since i don't intend to implement it now. I would like to complete working on this PR for now. So I removed PubSub.js now.

mccalluc commented on hubmap/frontend/src/middleware/PubSub.js:
```diff
@@ -0,0 +1,17 @@
+unsubscribe = (token) => {
+  PubSub.unsubscribe(token);
+}
+}
```
What is this for? The interface this offers is exactly the same as PubSub itself, so why is it necessary?

sushmaanandakoju replied:
We might want to add redux+pubsub at some point. but I removed it for now, since i don't intend to implement it now. I would like to complete working on this PR for now. So I removed PubSub.js now.

mccalluc commented on hubmap/frontend/src/middleware/index.js:
```diff
@@ -0,0 +1,7 @@
+
+export { default as rootReducer } from './rootReducer';
+export { default as store } from './store';
+export { default as PubSubApi } from './PubSub';
```
Is there a reason not to have the same name?

sushmaanandakoju replied:
We might want to add redux+pubsub at some point. but I removed it for now, since i don't intend to implement it now. I would like to complete working on this PR for now. So I removed PubSub.js now.

mccalluc commented on hubmap/frontend/src/middleware/index.js:
```diff
@@ -0,0 +1,7 @@
+/** *
```
```suggestion
/**
```
Please fix all occurrences, unless there's a reason to do this?

mccalluc commented on hubmap/frontend/src/middleware/index.js:
```diff
@@ -0,0 +1,7 @@
+/** *
+ * This exports all required components for middleware.
```
There are other files in this directory: What determines which ones go in the index.js, and and which don't?

sushmaanandakoju replied:
Added actions.js. Restructured reducers (moved them to middleware/reducers) since all reducers are accessed from rootReducer anyway.

mccalluc commented on hubmap/frontend/src/ui-components/AutoSuggest.js:
```diff
@@ -0,0 +1,208 @@
+import flatten from 'flat';
+import facets from '../data/searchfacets.json';
+/** *
+ * Example from: https://github.com/TeamWertarbyte/material-ui-chip-input/blob/master/stories/examples/react-autosuggest.js
```
Copying and pasting an entire file is not a good idea. Please take a moment to read an article or two on the problem. In this case, I can't tell that you understand how all these components work.

sushmaanandakoju replied:
I am sorry I did not understand what you mean here. We are directly using search facets json. so what is wrong with that ? or perhaps do you mean to just load the file at some other "commons" and just access the json object ?

mccalluc replied:
- The resulting UI is not coherent (four different search fields, etc.) -- I think you need to put more thought into the end result you want, and how to use these pieces to achieve that goal.
- Please do some reading on why copy-and-pasting is not considered a good practice.
- With any code that gets committed, another developer needs to be able to ask why you approached it one way and not another. When blocks like this are inserted, it makes it impossible to ask those questions or get good responses.

sushmaanandakoju replied:
@mccalluc Thanks for your comment. Please read my comments as follows and please clarify I understood your comment on this one correctly or not. If I did not understand: please do feel free to clarify in detail, I'd be glad to follow your suggestion.

- W.r.t _The resulting UI is not coherent (four different search fields, etc.) -- I think you need to put more thought into the end result you want, and how to use these pieces to achieve that goal._ 
As I mentioned in my original description for the PR, https://github.com/hubmapconsortium/hubmap-data-portal/pull/167#issue-321853427 ,
This is a W.i.p/known issue.  
> **Browse Summary must be readble, editable but should not look like SearchInput - W.I.P .**
**As per your suggestion and as per the description in PR: I agree, we should fix it as part of this PR. working on it. **

- W.r.t your comment about copying and pasting and asking me about my intent behind this : - 

1. My intent is to explore/experiment available react components from recommended material-ui supported components for auto-suggest/search suggestions and hence I provided specific references/source example when I used examples directly in my code here refer this commit : such as https://github.com/hubmapconsortium/hubmap-data-portal/blob/4cfccc2f8a77b471a242632e1b9fa52c575de0f8/hubmap/frontend/src/ui-components/AutoSuggest.js#L14 , where I have clearly stated that as a example.

1. I am asked to upload changes to github (whether I am testing/exploring components or whatever the temporary change might be). _This also definitely does **NOT** mean: we retain experimental code which is purely for exploration purpose._ So I am attempting to upload, document my progress on mockups, as I am exploring every day. 

1. Also, at this time I am not actively working on Auto Suggestion feature, so for the purpose of clarity and respecting your comment, without removing a record of my daily work, I have deleted experimental file AutoSuggest.js .

**As a takeaway from your comment, I am going to remove experimental/exploratory, temporary code/files before creating future PRs.**

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+  PaperProps: {
+    style: {
+      maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
+      // width: 300,
```
Why?

sushmaanandakoju replied:
removed commented code.

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+import * as Commons from '../commons';
+
+const ITEM_HEIGHT = 48;
+const ITEM_PADDING_TOP = 12;
```
Should these be used anywhere else? Why not just give a maxHeight?

sushmaanandakoju replied:
fixed

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+  },
+  MuiFormLabel: {
+    root: {
+      // color: grey[800],
```
Please do not check in commented out code without explanation... Do you anticipate uncommenting it?

sushmaanandakoju replied:
removed.

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+      },
+      '&:hover': {
+        borderColor: blue[500],
+        borderWidth: 1,
```
Repeated properties should be factored out:
```
        borderColor: blue[500],
        borderWidth: 1,
        fontColor: 'black',
```
When someone wants to change some aspect of the style, they should be able to make a change in one place, rather than searching the entire code base to find all the places it has been copied and pasted.

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+  },
+};
+// eslint-disable-next-line prefer-const
+let selectedMenuSummary = [];
```
This looks like a global variable that you're using to store application state? Please don't. Use react context, or redux, or plain object state, or anything.

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+export default class CustomSelectMenu extends React.Component {
+  menutoken = '';
+
+  pubsubObj = null;
```
- Please do not run words together without capitals: `menuToken`, `pubSubObj`.
- This might work, with the [right babel config](https://stackoverflow.com/questions/22528967/es6-class-variable-alternatives), but better in the constructor, particularly since you're just resetting it anyway.

sushmaanandakoju replied:
both are fixed.

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+    super(props);
+    this.state = {
+      menuitem: '',
+      menuname: '',
```
capitals

sushmaanandakoju replied:
fixed

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+    }
+    if (msg === Commons.UNCHECKED_MENU_OPTIONS
+        && selectedMenuSummary.indexOf(summary) === -1) {
+      const index = selectedMenuSummary.indexOf(summary);
```
move this to top of function and use instead of recalculating.

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+        && selectedMenuSummary.indexOf(summary) === -1) {
+      const index = selectedMenuSummary.indexOf(summary);
+      if (index > -1) {
+        selectedMenuSummary.splice(index, 1);
```
Is this code ever reached? The outer `if` asserts `index === -1`.

Commit:
add Test mock up of ui components to demo and review, for feedback on PR 167

Commit:
Add test mock up ui component in experiments

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+  }
+
+  render() {
+    const { menusections, menuname } = this.props;
```
capitals

sushmaanandakoju replied:
fixed

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+    const htmlElements = [];
+    const keys = [];
+    const values = [];
+    Object.keys(menusections).forEach((key) => {
```
`map` is preferable to `forEach`.

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+            this.setState({ ...this.state, menuitem: event.target.value });
+          }}
+          input={<Input id="select-multiple-placeholder" />}
+          // eslint-disable-next-line react/destructuring-assignment
```
Please follow style guide.

sushmaanandakoju replied:
fixed

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+          variant="outlined"
+        >
+          {htmlElements}
+
```
```suggestion
```

mccalluc commented on hubmap/frontend/src/ui-components/BaseSelectDropdown.js:
```diff
@@ -0,0 +1,143 @@
+  menuname: 'Menu',
+};
+CustomSelectMenu.propTypes = {
+  // eslint-disable-next-line react/forbid-prop-types
```
Is it not possible to be more precise? Explain why, if not.

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+/* eslint-disable react/prop-types */
```
If you don't believe we should follow this guideline, please make a separate PR for that.

sushmaanandakoju replied:
will do . this will be a different branch and PR and issue created to track:https://github.com/hubmapconsortium/hubmap-data-portal/issues/168

sushmaanandakoju replied:
this is actually auto-fixed, as there are no proptypes usage in this file.

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+        fontSize: '16px',
+      },
+    },
+    color: grey[800],
```
again, avoid copy and paste.

sushmaanandakoju replied:
auto-fixed due to other w.i.p changes w.r.t styling and converting this to more generic component.

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+
+export default class SelectedBrowseOptionsSummary extends React.Component {
+  summary = '';
+
```
```suggestion
```

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+  summary = '';
+
+  searchtoken = '';
+
```
```suggestion
```

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+  searchtoken = '';
+
+  menutoken = '';
+
```
```suggestion
```

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+
+  menutoken = '';
+
+  pubsubObj = null;
```
Same concerns as above: Why set these here, if you reset them in the constructor?

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+        chips: prevState.chips.filter((c) => c !== deletedChip),
+      }));
+    } else {
+      // alert('Why would you delete React?');
```
Why is the if/else needed?

sushmaanandakoju replied:
fixed.

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+      chips.push(summary);
+    }
+    this.setState({
+      // eslint-disable-next-line react/no-access-state-in-setstate
```
No: please do not ignore this warning.

sushmaanandakoju replied:
fixed : as the code is refactored with other ongoing changes as mentioned in PR. Also with that, all eslint disables were not required.

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+    });
+  }
+
+  componentWillMount() {
```
should also handle unmount.

sushmaanandakoju replied:
fixed.

sushmaanandakoju replied:
fixed

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+    }
+  }
+
+  selectedMenuSummaryAdded(msg, summary) {
```
This might be more clear split into two functions: Let the event routing handle this, instead of bothering with multiple `if`s.

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+        </div>
+      );
+    }
+    return (
```
- Do you actually need to return a value?
- Treat this as a "guard clause" and move it to the top.

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+              value={chips}
+              onDelete={(deletedChip) => this.handleDelete(deletedChip)}
+              onBlur={(event) => {
+                // eslint-disable-next-line react/destructuring-assignment
```
Again, please follow our style guide, or propose eliminating the rule if you don't think it's helpful.

sushmaanandakoju replied:
auto-fixed by other changes.

sushmaanandakoju replied:
this is auto-fixed by other changes w.r.t some w.i.p on this component outside this PR (as mentioned in PR description).

mccalluc commented on hubmap/frontend/src/ui-components/BrowseSummary.js:
```diff
@@ -0,0 +1,144 @@
+                }
+              }}
+              style={{
+                borderWidth: 0, '&:focus': { outline: 'none', borderWidth: 0 }, backgroundColor: grey[200], borderColor: grey[100], color: grey[500], textDecorationColor: grey[500],
```
set a variable above and use it here, rather than nesting. Wrap line. Avoid copy and paste.

sushmaanandakoju replied:
auto-fixed when implementing other changes from w.i.p.

mccalluc commented on hubmap/frontend/src/ui-components/ControlledChipInput.js:
```diff
@@ -0,0 +1,102 @@
+    this.pubsubObj = new PubSubApi();
+  }
+
+  onBeforeAdd = (chip) => chip.length >= 3
```
Is this right? I'm expected `onWhatever` to do something, instead of just returning a boolean...

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/ControlledChipInput.js:
```diff
@@ -0,0 +1,102 @@
+  },
+};
+
+const selectedValue = [];
```
again, please do not use global variables.

mccalluc commented on hubmap/frontend/src/ui-components/SubMenuOptions.js:
```diff
@@ -0,0 +1,118 @@
+    },
+  },
+})((props) => <Checkbox color="default" {...props} />);
+// eslint-disable-next-line no-var
```
again, no globals, and please follow our style guidelines.

mccalluc commented on hubmap/frontend/src/ui-components/SubMenuOptions.js:
```diff
@@ -0,0 +1,118 @@
+    }
+
+  handleChange = (menuname) => (event) => {
+    if (event.target.checked) {
```
`${menuname}:${event.target.value}` is copy and pasted four times.

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/SubMenuOptions.js:
```diff
@@ -0,0 +1,118 @@
+      checkedMenu: `${menuname}:${event.target.value}`,
+      checked: event.target.checked,
+    });
+    // event.target.checked = !event.target.checked;
```
no commented out code.

sushmaanandakoju replied:
removed.

sushmaanandakoju replied:
removed

mccalluc commented on hubmap/frontend/src/ui-components/SubMenuOptions.js:
```diff
@@ -0,0 +1,118 @@
+        >{menuname}
+        </FormLabel>
+        <FormGroup>
+          { menuitems ? menuitems.slice(0, menuitems.length).map((menuitem) => {
```
- Why do you copy menuitems with `menuitems.slice(0, menuitems.length)`?
- A ternary is hard to read beyond a line or two... probably better to ensure that menuitems is always at least an empty list, and then you don't need to worry about special cases.

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/SubMenuOptions.js:
```diff
@@ -0,0 +1,118 @@
+                    key={menuitem}
+                    checked={isChecked}
+                  />
+                    )}
```
fix indent.

sushmaanandakoju replied:
fixed.

mccalluc commented on hubmap/frontend/src/ui-components/index.js:
```diff
@@ -0,0 +1,9 @@
+/** *
```
Again, is there a reason why `ui-components/index.js` should just be focussed on the browse page?

sushmaanandakoju replied:
removed

mccalluc reviewed (changes requested):
Repeated code style problems:
- Capitalize each word after the first in every multi-word variable name.
- Do not copy and paste large chucks of code from the web: Understand what they do, and use only what you need.
- Do not copy and paste small chunks of code: factor it out into a variable, function, etc.
- Do not arbitrarily ignore eslint warnings.

In the UI:
- Fixed width makes it unusable at anything other than the widest window widths.
- There are four different components where I can put the cursor: Two "Search" boxes, and unlabelled line, and "Selection summary". How does the user know which one to type in?
- Checking and unchecking a box under "Assays" will add the same term multiple times to "Selection summary". 
- Closing a term in "Selection summary" does not uncheck the corresponding term under "Assays".

sushma-ananda committed:
Update hubmap/frontend/src/commons/index.js

Co-Authored-By: Chuck McCallum <mccalluc@users.noreply.github.com>

sushma-ananda committed:
Update hubmap/frontend/src/data/searchfacets.json

Co-Authored-By: Chuck McCallum <mccalluc@users.noreply.github.com>

sushma-ananda committed:
Update hubmap/frontend/src/middleware/PubSub.js

Co-Authored-By: Chuck McCallum <mccalluc@users.noreply.github.com>

sushma-ananda committed:
Update hubmap/frontend/src/middleware/index.js

Co-Authored-By: Chuck McCallum <mccalluc@users.noreply.github.com>

Commit:
undo adding this .directory

Commit:
remove /** * and just have /**

sushma-ananda committed:
Update hubmap/frontend/src/ui-components/BaseSelectDropdown.js

Co-Authored-By: Chuck McCallum <mccalluc@users.noreply.github.com>

sushma-ananda committed:
Update hubmap/frontend/src/ui-components/BrowseSummary.js

Co-Authored-By: Chuck McCallum <mccalluc@users.noreply.github.com>

sushma-ananda committed:
Update hubmap/frontend/src/ui-components/BrowseSummary.js

Co-Authored-By: Chuck McCallum <mccalluc@users.noreply.github.com>

sushma-ananda committed:
Update hubmap/frontend/src/ui-components/BrowseSummary.js

Co-Authored-By: Chuck McCallum <mccalluc@users.noreply.github.com>

Commit:
remove commented code

Commit:
add exceptions for props-spreading instead PR 167 (review)

Commit:
Merge branch 'browse-mockup-ui-components' of https://github.com/hubmapconsortium/hubmap-data-portal into browse-mockup-ui-components

Commit:
revert removing lines: agree with lines-between-class-members for travis build fix: #232

Commit:
remove repeated reference to selectedValue for menuName .

1) PR 167 (RC) :
https://github.com/hubmapconsortium/hubmap-data-portal/pull/167#discussion_r329207247

Commit:
fix indent

1) PR 167 RC:
https://github.com/hubmapconsortium/hubmap-data-portal/pull/167/files/fb82087fe09fa0b4552b52da3c703ecde9e20836#r329212649

mccalluc commented:
With something as complicated as this, some automated testing is necessary. Please look at [enzyme](https://github.com/airbnb/enzyme).

Commit:
rename data file

Commit:
change renamed filename

Commit:
change reference to renamed file

Commit:
add indentation:

1) PR #167
2) https://github.com/hubmapconsortium/hubmap-data-portal/pull/167#discussion_r329143872

Commit:
rename PubSubApi to PubSub

Commit:
remove PubSub.js

1) Earlier intent: to do some redux+pubsub code additionally here. Atleast is not required right now.
2) PR 167 RC
https://github.com/hubmapconsortium/hubmap-data-portal/pull/167#discussion_r329150477

Commit:
remove reference to PubSub file that has been deleted

1) refer : commit #22fe9104 for deleted file and PR 167

Commit:
update reference to PubSub

1) refer: commit #22fe9104

Commit:
update reference to PubSub

1) refer: commit #22fe9104

Commit:
update reference to PubSub

1) refer: commit #22fe9104

Commit:
update reference to PubSub

1) refer: commit #22fe9104

Commit:
remove old PubSub class contrustor code/references

Commit:
add actions.js to middleware exports

Commit:
update reference to actions

1) refer commit #3d6062fa

Commit:
update reference to actions

    1) refer commit #3d6062fa

Commit:
move all reducers to reducers folder

1) move all reducers to reducers folder
since all reducers are referenced via rootReducer anyway

Commit:
update reference to actions

        1) refer commit #3d6062fa

Commit:
update reference to actions

        1) refer commit #3d6062fa

Commit:
update reference to actions

        1) refer commit #3d6062fa

Commit:
update reference to actions

        1) refer commit #3d6062fa

Commit:
update reference to actions

        1) refer commit #3d6062fa

Commit:
update reference to actions

        1) refer commit #3d6062fa

Commit:
update reference to actions

        1) refer commit #3d6062fa

Commit:
remove AutoSuggest

1) not-to-be-final-mockup-but-only-fortesting
2) also not an active work-in-progress atm.

Commit:
remove SearchInput

1) not used in any TestUIComponent/anywhere

Commit:
remove exports of non-existent files

Commit:
remove references to experimental components that were removed

Commit:
add maxHeight instead, since it is a constant anyway.

Commit:
correct references to PubSub, remove references to old pubsub class

Commit:
eslint fix

Commit:
came case naming
PR 167

Commit:
camel case naming for variables

Commit:
update camel case for props, variables and update props references

Commit:
move variable decarations inside constructor

1) PR 167
https://github.com/hubmapconsortium/hubmap-data-portal/pull/167#discussion_r329172004

Commit:
remove unused object

Commit:
remove blank line

Commit:
remove blank line before constructor

Commit:
[W.I.P]changes for BrowseSummary (to it uneditable input + PR 167 RC)

1) make BrowseSummary only as a readable component
2) remove ChipInput, instead use material-ui chip
3) remove unused imports
4) move declarations to cosntructor

Commit:
remove eslint disable for destructuring (not valid anymore), if/else block

1) remove if/else block
2) eslint disable for destructuring (not valid anymore)

Commit:
add unmount to unsubscribe, calculate once

PR 167 RC

Commit:
remove onBeforeAdd

Commit:
add separate events for each topic subscriptions

PR 167 RC

Commit:
move all ui-components' styles, theme HOCs to styles.js
PR 167 RC

Commit:
update references to select dropdown theme and MenuProps from styles.js

Commit:
update all references to controlled chip theme to styles.js

Commit:
update all references to Greycheckbox, useStyles to styles.js

Commit:
replace forEach with map
PR 167 RC

Commit:
rename to onMenuOptionsChanged and remove repeated logic.

mccalluc commented:
When you enter multiple terms, the behavior is weird:
- Type "a", hit return: You get () (a)
- "b", hit return: (a) (b)
- "c": (ab) (c)
- "d": (abc) (d)

You never get more than two tags, and the body of the second tag is appended to the first with successive entries.

I'm not going to spend more time on this now. At this point we need to pivot, and simplify the UI to what can be supported by the API, so I'm not sure if it's the best use of your time to continue with this.

mccalluc reviewed (commented):
Still buggy; have not reviewed code.

Commit:
[w.i.p] : checkbox state exists, just save in parent state
PR 167 RC

Commit:
add variables instead of hardcoded values for checkbox color

Commit:
add variables instead of hardcoded values for checkbox color

sushma-ananda commented:
> When you enter multiple terms, the behavior is weird:
> 
> * Type "a", hit return: You get () (a)
> * "b", hit return: (a) (b)
> * "c": (ab) (c)
> * "d": (abc) (d)
> 
> You never get more than two tags, and the body of the second tag is appended to the first with successive entries.
> 
> I'm not going to spend more time on this now. At this point we need to pivot, and simplify the UI to what can be supported by the API, so I'm not sure if it's the best use of your time to continue with this.

Hey much appreciate you checking status on this PR. This is w.i.p as discussed in today's call. please do hold on until I inform you that this is done and ready for you to test.

sushmaanandakoju commented:
> Still buggy; have not reviewed code.

Hey much appreciate you checking status on this PR. Well, is w.i.p as discussed in today's call. please do hold on until I inform you that this is done and ready for you to test. thanks!

Commit:
Merge branch 'browse-mockup-ui-components' of https://github.com/hubmapconsortium/hubmap-data-portal into browse-mockup-ui-components

Commit:
remove global variable and all file-level eslint

1) instead maintain a lit of menuitems
2) pass on the selected menu summary to child dropdown
3) remove forbidden proptypes eslint
4) remove eslint-destructuring assignment

Commit:
remove global variable and just publish changes

1) let parent send the props
2) publish selected options
3) remove eslint forbidden proptypes
4) remove eslint for unused variables
